msprime/intervals.py

```diff
@@ -229,26 +342,104 @@ def uniform_map(cls, length, rate, num_loci=None):
+        Parses the specified file in HapMap format. These files must be
+        white-space-delimited, and contain a single header line (which is
+        ignored), and then each subsequent line contains the starting position
+        and recombination rate for the segment from that position (inclusive)
+        to the starting position on the next line (exclusive). Starting
+        positions of each segment are given in units of bases, and
+        recombination rates in centimorgans/Megabase. The first column in this
+        file is checked , as are additional columns after the third (Position is
+        assumed to be the second column, and Rate is assumed to be the third).
+        If the first starting position is not equal to zero, then a
+        zero-recombination region is inserted at the start of the chromosome.
 
         .. warning::
             This method is deprecated, use the module-level
-            :func:`read_hapmap` function instead.
+            :func:`read_hapmap` function instead, which uses the genetic position
+            in centiMorgans rather than the rate column.
 
         :param str filename: The name of the file to be parsed. This may be
-            in plain text or gzipped plain text.
+            in plain text or gzipped plain text. It may also be a file object.
+        :param float sequence_length: The .
         :return: A RecombinationMap object.
         """
         warnings.warn(
             "RecombinationMap.read_hapmap() is deprecated. "
             "Use msprime.read_hapmap() instead.",
             FutureWarning,
         )
-        rate_map = read_hapmap(filename)
-        rate = np.append(rate_map.rate, 0)
-        return cls(rate_map.position, rate, map_start=rate_map.map_start)
+        hapmap = np.loadtxt(filename, skiprows=1, usecols=(1, 2))
+        try:
+            filename.seek(0)  # If a file object, we need to read from start again
+        except AttributeError:
+            pass
+        rownames = np.genfromtxt(filename, skip_header=True, usecols=0, dtype="str")
+        if len(np.unique(rownames)) > 1:
+            warnings.warn(
+                f"Different chromosome names found in the first column of {filename}. "
+                "Check that the file contains data for a single chromosome."
+            )
+        position = hapmap[:, 0]
+        # Rate is expressed in centimorgans per megabase, which
+        # we convert to per-base rates
+        rate = 1e-8 * hapmap[:, 1]
+
+        map_start = position[0]
+        map_end = position[-1]
+        if map_start != 0:
+            position = np.insert(position, 0, 0)
+            rate = np.insert(rate, 0, 0)
+        if rate[-1] != 0:
+            raise ValueError(
+                "The last rate provided in the recombination map must be zero"
+            )
+        if sequence_length is not None:
+            position = np.append(position, sequence_length)
+            rate = np.append(rate, 0)
```

Comment:
I think this has three cases that need to be handled differently. 1) `if sequnce_length < position[-1]`, raise an error, 2) `elif sequence_length == position[-1]`, do nothing. 3) `else:` append as you have here.

Reply:
Good point. I haven't done the unit tests for sequence_length yet.